equistore-core/src/blocks.rs

```diff
@@ -104,152 +95,91 @@ fn check_component_labels(components: &[Arc<Labels>]) -> Result<(), Error> {
-                let mut property = new_property.to_vec();
-                property.extend_from_slice(old_property);
-                new_properties_builder.add(&property)?;
-            }
-        }
-        let new_properties = new_properties_builder.finish();
-
-        let mut new_shape = self.data.shape()?.to_vec();
-        let properties_axis = new_shape.len() - 1;
-        new_shape[properties_axis] = new_properties.count();
-        new_shape.remove(component_axis + 1);
-
-        self.data.swap_axes(component_axis + 1, properties_axis - 1)?;
-        self.data.reshape(&new_shape)?;
-
-        self.properties = Arc::new(new_properties);
-
-        Ok(())
-    }
-
-    /// Try to copy this `BasicBlock`. This can fail if we are unable to copy
-    /// the underlying `eqs_array_t` data array
-    pub fn try_clone(&self) -> Result<BasicBlock, Error> {
-        let data = self.data.try_clone()?;
-
-        Ok(BasicBlock {
-            data,
-            samples: Arc::clone(&self.samples),
-            components: self.components.clone(),
-            properties: Arc::clone(&self.properties),
-        })
-    }
-}
-
 /// A single block in a `TensorMap`, containing both values & optionally
 /// gradients of these values w.r.t. any relevant quantity.
 #[derive(Debug)]
 pub struct TensorBlock {
-    values: BasicBlock,
-    gradients: HashMap<String, BasicBlock>,
+    pub values: eqs_array_t,
+    pub samples: Arc<Labels>,
+    pub components: ImmutableVec<Arc<Labels>>,
+    pub properties: Arc<Labels>,
+    gradients: HashMap<String, TensorBlock>,
     // all the keys from `self.gradients`, as C-compatible strings
     gradient_parameters: Vec<ConstCString>,
 }
 
 impl TensorBlock {
-    /// Create a new `TensorBlock` containing the given data, described by the
+    /// Create a new `TensorBlock` containing the given values, described by the
     /// `samples`, `components`, and `properties` labels. The block is
     /// initialized without any gradients.
     pub fn new(
-        data: impl Into<eqs_array_t>,
+        values: eqs_array_t,
         samples: Arc<Labels>,
         components: Vec<Arc<Labels>>,
         properties: Arc<Labels>,
     ) -> Result<TensorBlock, Error> {
+        check_values_and_labels(
+            "values and labels don't match", &values, &samples, &components, &properties
+        )?;
+        check_component_labels(&components)?;
+        let components = ImmutableVec(components);
         Ok(TensorBlock {
-            values: BasicBlock::new(data.into(), samples, components, properties)?,
+            values, 
+            samples, 
+            components, 
+            properties,
             gradients: HashMap::new(),
             gradient_parameters: Vec::new(),
         })
     }
 
     /// Try to copy this `TensorBlock`. This can fail if we are unable to copy
-    /// one of the underlying `eqs_array_t` data arrays
+    /// one of the underlying `eqs_array_t` values arrays
```

Comment:
Same here, the doc could keep using data, and the data is passed in the `values` argument

Reply:
I've tried going through all the comments but, in some cases, it is difficult to tell whether I should be using values or data... simply because they're fundamentally the same thing. Should we rename everything to data (or to values)? In case we go for data, it will be a massive breaking change, but I think this is the correct thing to do in the long run

Reply:
So in my mind "values" in in opposition to "gradients", while "data" is in opposition to "metadata". I can help going through the code to find the best one, but I don't think we can use only one of them everywhere here

Reply:
Ok, in that case I'll try my best and then you'll check